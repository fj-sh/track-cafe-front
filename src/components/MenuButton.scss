.order-button {
  width: 300px;
  position: relative;
  border-radius: 4px;
  border: 2px solid #262626;
  padding: 15px 30px;
  color: #101010;
  cursor: pointer;
  user-select: none;
  transition: all .3s;
  box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;

  &:hover {
    transform: scale(1.03);
    background: rgba(0,0,0,.13);
    box-shadow: 0 2px 20px rgba(0,0,0,.15);
  }

  &:active {
    transform: scale(.96);

    .badge {
      transform: scale(1.2);
    }
  }
}

.menu {
  display: flex;
  justify-content: space-between;
}

.badge {
  border-radius: 50%;
  color: #f5f5f5;
  width: 40px;
  height: 40px;
  display: block;
  position: absolute;
  background: #f8011a;
  align-items: center;
  box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
  justify-content: center;
  top: -15px;
  right: -15px;
  transition: all .3s;
}
