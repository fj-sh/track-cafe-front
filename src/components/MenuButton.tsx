import React from 'react'
import { Menu } from '../lib/types/menu'
import style from './MenuButton.module.scss'

type Props = {
  menu: Menu
  updateMenus: (menu: Menu) => void
}

const MenuButton = ({ menu, updateMenus }: Props) => {
  const onClick = () => {
    updateMenus(menu)
  }
  return (
    <>
      <div className={style.menuButton} onClick={onClick} data-testid="menuButton">
        <div className={style.menu}>
          <span className="">{menu.name}</span>
          <span className="">{menu.price} å††</span>
        </div>
        <span className={style.badge} data-testid="badge">
          {menu.numberOfOrders}
        </span>
      </div>
    </>
  )
}

export default MenuButton
