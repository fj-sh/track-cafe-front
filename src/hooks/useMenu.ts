import { useState } from 'react'
import { Menu } from '../lib/types/menu'

/**
 * 初期メニューからメニューの状態を生成する。
 * @param {Menu[]} initialMenu
 * @return {Object} メニューのステート
 */
export const useMenu = (initialMenu: Menu[]) => {
  const [menus, setMenus] = useState(initialMenu)
  const updateMenus = (targetMenu: Menu) => {
    const updatedMenus = menus.map((menu) => {
      if (menu.id === targetMenu.id) {
        return { ...menu, numberOfOrders: menu.numberOfOrders + 1 }
      }
      return menu
    })
    setMenus(updatedMenus)
  }
  return {
    menus,
    updateMenus,
  }
}
