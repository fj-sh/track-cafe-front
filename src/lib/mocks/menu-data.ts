import { Menu } from '../types/menu'

export const menuData: Menu[] = [
  { id: 'coffee', name: 'コーヒー', price: 480, numberOfOrders: 0 },
  { id: 'tea', name: '紅茶', price: 280, numberOfOrders: 0 },
  { id: 'milk', name: 'ミルク', price: 180, numberOfOrders: 0 },
  { id: 'coke', name: 'コーラ', price: 190, numberOfOrders: 0 },
  { id: 'beer', name: 'ビール', price: 580, numberOfOrders: 0 },
]
