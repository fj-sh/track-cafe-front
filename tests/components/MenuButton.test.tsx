import React from 'react'
import { vi } from 'vitest'
import { cleanup, fireEvent, render, renderHook } from '@testing-library/react'
import { screen } from '@testing-library/react'
import MenuButton from '../../src/components/MenuButton'
import { useMenu } from '../../src/hooks/useMenu'
import { menuData } from '../../src/lib/mocks/menu-data'

describe('MenuButton', () => {
  it('正しくレンダリングされること', () => {
    const { result } = renderHook(() => useMenu(menuData))
    const { container } = render(
      <MenuButton menu={result.current.menus[0]} updateMenus={result.current.updateMenus} />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  describe('menuButtonがクリックされたとき', () => {
    const handleClick = vi.fn()
    beforeAll(() => {
      const { result } = renderHook(() => useMenu(menuData))
      render(<MenuButton menu={result.current.menus[0]} updateMenus={handleClick} />)
      fireEvent.click(screen.getByTestId('menuButton'))
    })
    it('onClickが呼ばれること', () => {
      expect(handleClick).toHaveBeenCalledTimes(1)
    })
  })
  afterEach(() => cleanup())
})
