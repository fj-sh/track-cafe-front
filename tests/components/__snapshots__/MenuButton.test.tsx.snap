// Vitest Snapshot v1

exports[`MenuButton > 正しくレンダリングされること 1`] = `
<div
  class="_menuButton_fdi0z_1"
  data-testid="menuButton"
>
  <div
    class="_menu_fdi0z_1"
  >
    <span
      class=""
    >
      コーヒー
    </span>
    <span
      class=""
    >
      480
       円
    </span>
  </div>
  <span
    class="_badge_fdi0z_21"
    data-testid="badge"
  >
    0
  </span>
</div>
`;
